### DISCUSSION TITLE
SDK access (tasklib or equivalent?)

### DISCUSSION THREAD

**Discussion by echo-bravo-yahoo** [GothenburgBitFactory/taskwarrior#3476] [General]
#### SDK access
The rewrite to 3.0.0 seems like a good opportunity to explore programmatic access that aims to expand the range of platforms taskwarrior is easily usable on. For instance, android and iphone support has been a problem for taskwarrior users for a while, with very few, very barebones clients available. How can 3.0.0 better support developers building interfaces in environments that taskwarrior was previously difficult to build on?

##### A little more context
I've been using Obsidian a lot, and wanted to have easy access to tasks. On desktop computers, this is quite easy - install taskwarrior's binary, and use nodeJS' child_process module to run it from the CLI and parse the result. On mobile, however, this strategy doesn't work at all. Ideally, moving to Rust might make it easier to publish first-party SDKs or bindings to more languages, with the only dependency being that SDK and a local SQLite DB file.

**Comment by djmitche** [comment]
This was part of the motivation for the changes in 3.0. The idea is that TaskChampion allows programmatic access to tasks. That can either run against the same replica as used by the CLI or use the sync system to sync changes into the CLI. The latter is best for situations like a mobile app, web app, or a gateway like bugwarrior - they can make changes to their local replica and rely on sync to communicate those changes to other replicas.

We've lost a bit of momentum on TaskChampion with the integration into TaskWarrior, and the language interface situation isn't great -- there's a `taskchampion-lib` that probably should be deleted. There's a new [Python API under construction](https://github.com/GothenburgBitFactory/taskchampion/pull/388), though, which should open a lot of doors! I think we can get to a place where there are decent interfaces to Taskchampion in most popular higher-level languages -- it just needs more work.

**Comment by wrashi** [comment]
I'm also a heavy Obsidian user and would love to have integration like vimwiki/taskwiki. These video demonstrations make it look pretty magical, so very excited about the TaskChampion integration.
https://www.youtube.com/watch?v=A1YgbAp5YRc
https://www.youtube.com/watch?v=UuHJloiDErM
